File: C#/125. Valid Palindrome.cs
public class Solution
{
    bool IsLegal(char ch)
    {
        return (ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9') || (ch >= 'A' && ch <= 'Z');
    }
    public bool IsPalindrome(string s)
    {
        int lPos = 0, rPos = s.Length - 1;
        while (lPos < rPos)
        {
            while (lPos < s.Length && !IsLegal(s[lPos]))
                ++lPos;
            while (rPos >= 0 && !IsLegal(s[rPos]))
                --rPos;
            if (lPos >= rPos)
                break;
            if (Char.ToLower(s[lPos++]) != Char.ToLower(s[rPos--]))
                return false;
        }
        return true;
    }
}
//
public class Solution
{
    bool IsLegal(char ch)
    {
        return (ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9') || (ch >= 'A' && ch <= 'Z');
    }
    public bool IsPalindrome(string s)
    {
        int lPos = 0, rPos = s.Length - 1;
        while (lPos < rPos)
        {
            while (lPos < s.Length && !IsLegal(s[lPos]))
                ++lPos;
            while (rPos >= 0 && !IsLegal(s[rPos]))
                --rPos;
            if (lPos >= rPos)
                break;
            if ((s[lPos] == s[rPos]) || (s[rPos] >= 'a' && s[rPos] <= 'z' && s[lPos] == s[rPos] + 'A' - 'a') || (s[rPos] >= 'A' && s[rPos] <= 'Z' && s[lPos] == s[rPos] - 'A' + 'a'))
            {
                ++lPos;
                --rPos;
                continue;
            }
            return false;
        }
        return true;
    }
}

File: python/125. Valid Palindrome.py
# -*- coding: utf-8 -*-

"""
@Author: xiezizhe
@Date: 28/2/2020 上午2:41
"""


class Solution:
    def valid(self, c):
        return ('z' >= c >= 'a') or ('9' >= c >= '0')

    def isPalindrome(self, s: str) -> bool:
        if s is None or s == '' or len(s) == 1:
            return True

        idx1, idx2 = 0, len(s) - 1
        s = s.lower()
        while idx1 < idx2:
            while idx1 < idx2 and not self.valid(s[idx1]):
                idx1 += 1
            while idx1 < idx2 and not self.valid(s[idx2]):
                idx2 -= 1

            if s[idx1] != s[idx2]:
                return False
            idx1 += 1
            idx2 -= 1
        return True


if __name__ == "__main__":
    s = Solution()
    print(s.isPalindrome('5 6 6 $# 7 5'))
